src/derive-c/allocs/test/template.h

```diff
@@ -0,0 +1,155 @@
+/// @brief For unit tests expected to throw, as C has no unwind, we cannot free allocated memory.
+///        This macro wraps the allocator in debug, to allow clearing leaks after an exception.
+///
+/// In release, it is a no-op / pass through.
+///
+/// As this is entirely C, we do not get the niceties of a C++ RAII allocator guard shebang.
+// However, this is usable inside unit tests written in C.
+
+#include <stddef.h>
+
+#include <derive-c/allocs/std.h>
+#include <derive-c/core.h>
+#include <derive-c/panic.h>
+#include <derive-c/self.h>
+
+#ifndef ALLOC
+#ifndef __clang_analyzer__
+#error "The allocator being wrapped must be defined"
+#endif
+#include <derive-c/allocs/null.h> // NOLINT(misc-include-cleaner)
+#define ALLOC nullalloc
+#endif
+
+#ifndef ENTRIES
+#ifndef __clang_analyzer__
+#error "The name of the type to generate for storing entries must be defined"
+#endif
+#define ENTRIES derive_c_entries_placeholder_name
+#endif
+
+#ifdef NDEBUG
+
+typedef struct {
+    ALLOC* alloc;
+} SELF;
+
+static SELF NAME(SELF, new)(ALLOC* alloc) { return (SELF){.alloc = alloc}; }
+
+static void* NAME(SELF, malloc)(SELF* self, size_t size) {
+    return NAME(ALLOC, malloc)(self->alloc, size);
+}
+
+static void* NAME(SELF, calloc)(SELF* self, size_t count, size_t size) {
+    return NAME(ALLOC, calloc)(self->alloc, count, size);
+}
+
+static void* NAME(SELF, realloc)(SELF* self, void* ptr, size_t size) {
+    return NAME(ALLOC, realloc)(self->alloc, ptr, size);
+}
+
+static void NAME(SELF, free)(SELF* self, void* ptr) { return NAME(ALLOC, free)(self->alloc, ptr); }
+
+#else
+
+#define TRACKED_ENTRY NAME(ENTRIES, entry)
+
+typedef struct {
+    void* ptr;
+    bool freed;
+} TRACKED_ENTRY;
+
+#pragma push_macro("SELF")
+#undef SELF
+
+// JUSTIFY: Using a vector rather than a faster lookup map.
+//           - Give this will be used for test & debug, performance matters less
+//           - Much easier to explore a vector, than a hashmap in gdb.
+#define T TRACKED_ENTRY
+#define ALLOC stdalloc
+#define SELF ENTRIES
+#include <derive-c/structures/vector/template.h>
+
+#pragma pop_macro("SELF")
```

Comment:
pattern is required to make use of templated structures in templated structures - cannot temporarily redefine `SELF` using previous `SELF`, so need a whole extra input (`ENTRIES`)